analyzers/src/SonarAnalyzer.Common/Rules/Utilities/SymbolReferenceAnalyzerBase.cs

```diff
@@ -79,86 +62,85 @@ protected sealed override SymbolReferenceInfo CreateMessage(SyntaxTree syntaxTre
+                }
             }
-            return sr;
+
+            return references;
         }
 
-        private SymRefInfo GetSymRefInfo(SyntaxToken token, SemanticModel semanticModel)
+        private static ISymbol GetReferenceSymbol(SyntaxNode node, SemanticModel model) =>
+            model.GetSymbolInfo(node).Symbol switch
+            {
+                IMethodSymbol { MethodKind: MethodKind.Constructor, IsImplicitlyDeclared: true } constructor => constructor.ContainingType,
+                var symbol => symbol
+            };
+
+        private static SymbolReferenceInfo.Types.SymbolReference GetSymbolReference(ReferenceInfo[] references, SyntaxTree tree)
         {
-            if (!Language.Syntax.IsKind(token, Language.SyntaxKind.IdentifierToken))
+            var declarationSpan = GetDeclarationSpan(references);
+            if (!declarationSpan.HasValue)
             {
-                // For the time being, we only handle identifier tokens.
-                // We could also handle keywords, such as this, base
                 return null;
             }
 
-            if (semanticModel.GetDeclaredSymbol(token.Parent) is { } declaredSymbol)
+            var symbolReference = new SymbolReferenceInfo.Types.SymbolReference { Declaration = GetTextRange(Location.Create(tree, declarationSpan.Value).GetLineSpan()) };
+            foreach (var reference in references.Where(x => !x.IsDeclaration).Select(x => x.Identifier))
             {
-                return declarationKinds.Contains(declaredSymbol.Kind)
-                    ? new SymRefInfo(token, declaredSymbol, true)
-                    : null;
+                symbolReference.Reference.Add(GetTextRange(Location.Create(tree, reference.Span).GetLineSpan()));
             }
-
-            if (GetBindableParent(token) is { } node)
-            {
-                var symbol = semanticModel.GetSymbolInfo(node).Symbol;
-                if (symbol == null)
-                {
-                    return null;
-                }
-                else if (symbol.DeclaringSyntaxReferences.Any() || IsValuePropertyParameter(symbol))
-                {
-                    return new SymRefInfo(token, symbol);
-                }
-                else if (symbol is IMethodSymbol ctorSymbol && ctorSymbol.MethodKind == MethodKind.Constructor && ctorSymbol.IsImplicitlyDeclared)
-                {
-                    return new SymRefInfo(token, ctorSymbol.ContainingType);
-                }
-            }
-
-            return null;
+            return symbolReference;
         }
 
+        private static TextSpan? GetDeclarationSpan(IEnumerable<ReferenceInfo> references) =>
+            references.FirstOrDefault(x => x.IsDeclaration) is { } declaration
+                ? declaration.Identifier.Span
+                : null;
+
         private static bool HasTooManyTokens(SyntaxTree syntaxTree) =>
             syntaxTree.GetRoot().DescendantTokens().Count() > TokenCountThreshold;
```

Comment:
More of a question, if we only process `DescendantTokens` of  kind `IndentifierToken`, should we also check here only tokens of that type when comparing to the treshold?

Reply:
See also `GetDeclarations`. We visit the entire tree and retrieve symbols for all the declarations defined there. I would keep this unchanged for now.